### ISSUE TITLE
Exception when downloading Kraken trades

### ISSUE THREAD

**Issue by hroff-1902** [freqtrade/freqtrade#3404]
## Describe your environment

  * Operating system: Ubuntu
  * Python Version: 3,7.7 (`python -V`)
  * CCXT version: 1.28.49 (`pip freeze | grep ccxt`)
  * Freqtrade Version: develop (`freqtrade -V` or `docker-compose run --rm freqtrade -V` for Freqtrade running in docker)
  
## Describe the problem:

Have `download-data --dl-trades` running for Kraken, with many retries, finally got exception:
```
2020-05-30 19:43:45,991 - freqtrade.exchange.exchange - DEBUG - Fetching trades for pair ZEC/EUR, params: {'since': '1590856217466086643'} 
2020-05-30 19:43:49,020 - freqtrade.exchange.exchange - DEBUG - Stopping because from_id did not change. Reached 1590856953714 > 1590856750421
2020-05-30 19:43:49,057 - freqtrade.misc - INFO - dumping json to "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/user_data/data/kraken/ZEC_EUR-trades.json.gz"
2020-05-30 19:43:49,840 - freqtrade.misc - DEBUG - done json to "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/user_data/data/kraken/ZEC_EUR-trades.json.gz"
2020-05-30 19:43:49,841 - freqtrade.data.history.history_utils - DEBUG - New Start: 2020-02-20T03:00:04
2020-05-30 19:43:49,841 - freqtrade.data.history.history_utils - DEBUG - New End: 2020-05-30T19:42:33
2020-05-30 19:43:49,841 - freqtrade.data.history.history_utils - INFO - New Amount of trades: 68966
2020-05-30 19:43:49,868 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/main.py", line 36, in main
    return_code = args['func'](args)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/commands/data_commands.py", line 61, in start_download_data
    data_format_trades=config['dataformat_trades'],
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/data/history/history_utils.py", line 352, in convert_trades_to_ohlcv
    ohlcv = trades_to_ohlcv(trades, timeframe)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/data/converter.py", line 200, in trades_to_ohlcv
    return df_new[DEFAULT_DATAFRAME_COLUMNS]
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/.env/lib/python3.7/site-packages/pandas/core/series.py", line 910, in __getitem__
    return self._get_with(key)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/.env/lib/python3.7/site-packages/pandas/core/series.py", line 958, in _get_with
    return self.loc[key]
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/.env/lib/python3.7/site-packages/pandas/core/indexing.py", line 1768, in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/.env/lib/python3.7/site-packages/pandas/core/indexing.py", line 1954, in _getitem_axis
    return self._getitem_iterable(key, axis=axis)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/.env/lib/python3.7/site-packages/pandas/core/indexing.py", line 1595, in _getitem_iterable
    keyarr, indexer = self._get_listlike_indexer(key, axis, raise_missing=False)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/.env/lib/python3.7/site-packages/pandas/core/indexing.py", line 1553, in _get_listlike_indexer
    keyarr, indexer, o._get_axis_number(axis), raise_missing=raise_missing
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/.env/lib/python3.7/site-packages/pandas/core/indexing.py", line 1655, in _validate_read_indexer
    "Passing list-likes to .loc or [] with any missing labels "
KeyError: 'Passing list-likes to .loc or [] with any missing labels is no longer supported, see https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike'
2020-05-30 19:43:49,871 - freqtrade.exchange.exchange - DEBUG - Exchange object destroyed, closing async loop
```

**Comment by hroff-1902** [comment]
seems it's introduced by pandas 1.0.4 (I've updated pandas)

this probably should be fixed before master release and included in it

**Comment by xmatthias** [comment]
i very strongly doubt that it's been introduced between 1.0.3 and 1.0.4 - but rather in 1.0.0 (if at all).

However, i'm unable to recreate this (just ran a full download of a few pairs (not too many - as rate-limiting makes kraken very slow) - which makes me suspect that something in your configuration is off (you mention manu retries - which suggests that your rateLimit is not correctly set).

I'll not consider this a stopper for a release unless we're able to reproduce this with accuracy.

**Comment by xmatthias** [comment]
I suspect this can happen if the dataframe is empty - so most likely, you've got one trades file which is empty basically contains (`[]`).

we can fix that for sure - but it's not critical - as it's been there in the last few releases too (empty trades data never did well...).

**Comment by xmatthias** [comment]
Well the fix to this issue is easy - change line 200 of converter from 

``` python
    return df_new[DEFAULT_DATAFRAME_COLUMNS]
```
to 
```python
    return df_new.loc[:, DEFAULT_DATAFRAME_COLUMNS]
```

The problem is - the error you're seeing is just a symptom (dataframe is empty - i assume) - so i'd really get to analyze that first (but as i'm not experiencing it - it's difficult for me ...).

**Comment by laukaho** [comment]
I am using docker-compose and this error is not fixed

**Comment by xmatthias** [comment]
@hroff-1902 can we close this, since #3444 and #3432 do address this (one by fixing the crash, and one by explaining how to configure for downloads) .. ?

**Comment by hroff-1902** [comment]
well, I have a couple of more issues to fix regarding this, I would like to open it for some time as a reminder to avoid creation of new issues...

the points are:
* better debug logging
* an option to control the number of retries for exchange operation; there was something with my network connection during those 2 days, and I was only able to download trades by increasing the number of retries (experienced rows of Reset Connection errors up to 15 errors with rateLimit=2000 and 3100)

**Comment by xmatthias** [comment]
there's a difference i think - connectionTimeout (networkProblem) is not the same as a DDOS error (where the exchange blocks you).

You can set rateLimit to 100k if you're seeing connectionErrors / ConnectionTimeout - that's not what rateLimit is or (kraken/ccxt does give you an explicit DDosException if that's what happened).

**Comment by Blackhawke** [comment]
As I'm a noob to FreqTrade (but have been running linux since 1996) I fear that posting this might either be unhelpful or add to the confusion, but I'm getting the API rate-limit error even when asking for list-pairs, or test-pairlist. Not just when running "download-data --dl-trades". The last command I ran was "list-pairs". It returned the following (log level 3):

`
2020-06-28 12:33:37,106 - freqtrade.configuration.configuration - INFO - Using pairlist from configuration.
2020-06-28 12:33:37,106 - freqtrade.configuration.config_validation - INFO - Validating configuration ...
2020-06-28 12:33:37,110 - freqtrade.exchange.exchange - INFO - Instance is running with dry_run enabled
2020-06-28 12:33:37,110 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True}
2020-06-28 12:33:37,113 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True, 'rateLimit': 3100}
2020-06-28 12:33:37,114 - asyncio - DEBUG - Using selector: EpollSelector
2020-06-28 12:33:37,118 - freqtrade.exchange.exchange - INFO - Using Exchange "Kraken"
2020-06-28 12:33:37,118 - freqtrade.resolvers.exchange_resolver - INFO - Using resolved exchange 'Kraken'...
2020-06-28 12:33:37,119 - freqtrade.exchange.exchange - WARNING - Markets were not loaded. Loading them now..
2020-06-28 12:33:37,119 - ccxt.base.exchange - DEBUG - GET https://api.kraken.com/0/public/Assets, Request: {'User-Agent': 'python-requests/2.23.0', 'Accept-Encoding': 'gzip, deflate'} None
2020-06-28 12:33:37,122 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): api.kraken.com:443
2020-06-28 12:33:37,306 - urllib3.connectionpool - DEBUG - https://api.kraken.com:443 "GET /0/public/Assets HTTP/1.1" 200 None
2020-06-28 12:33:37,342 - ccxt.base.exchange - DEBUG - GET https://api.kraken.com/0/public/Assets, Response: 200 {'Date': 'Sun, 28 Jun 2020 19:33:37 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Set-Cookie': '__cfduid=d8eadf8f230f9d42354a57163338a23831593372817; expires=Tue, 28-Jul-20 19:33:37 GMT; path=/; domain=.kraken.com; HttpOnly; SameSite=Lax', 'cache-control': 'public,max-age=29,smaxage=29', 'content-encoding': 'gzip', 'vary': 'Accept-Encoding', 'strict-transport-security': 'max-age=15768000', 'CF-Cache-Status': 'HIT', 'Age': '2', 'cf-request-id': '039e04cf7c0000ed6b4305d200000001', 'Expect-CT': 'max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"', 'Server': 'cloudflare', 'CF-RAY': '5aa9d72bf949ed6b-SJC'} {"error":[],"result":{"ADA":{"aclass":"currency",

(=snip long list of currency pairs to EOL=)

2020-06-28 12:33:40,238 - ccxt.base.exchange - DEBUG - GET https://api.kraken.com/0/public/AssetPairs, Response: 200 {'Date': 'Sun, 28 Jun 2020 19:33:40 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Set-Cookie': '__cfduid=d8ff9a690e2572fb56afd8c24edd90c5f1593372820; expires=Tue, 28-Jul-20 19:33:40 GMT; path=/; domain=.kraken.com; HttpOnly; SameSite=Lax', 'cache-control': 'public,max-age=45,smaxage=45', 'content-encoding': 'gzip', 'vary': 'Accept-Encoding', 'strict-transport-security': 'max-age=15768000', 'CF-Cache-Status': 'HIT', 'Age': '22', 'cf-request-id': '039e04dab00000ed6b430b8200000001', 'Expect-CT': 'max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"', 'Server': 'cloudflare', 'CF-RAY': '5aa9d73dea41ed6b-SJC'} {"error":[],"result":{"ADAETH":{"altname":"ADAETH","wsname":"ADA\/ETH","aclass_base":"currency","base":"ADA","aclass_quote":"currency",

(=snip another long list of currency pairs to EOL=)

2020-06-28 12:33:43,125 - ccxt.base.exchange - DEBUG - GET https://kraken.zendesk.com/api/v2/help_center/en-us/articles/205893708, Request: {'User-Agent': 'python-requests/2.23.0', 'Accept-Encoding': 'gzip, deflate'} None
2020-06-28 12:33:43,126 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): kraken.zendesk.com:443
2020-06-28 12:33:43,596 - urllib3.connectionpool - DEBUG - https://kraken.zendesk.com:443 "GET /api/v2/help_center/en-us/articles/205893708 HTTP/1.1" 429 50
2020-06-28 12:33:43,599 - ccxt.base.exchange - DEBUG - GET https://kraken.zendesk.com/api/v2/help_center/en-us/articles/205893708, Response: 429 {'Date': 'Sun, 28 Jun 2020 19:33:43 GMT', 'Content-Type': 'text/plain', 'Content-Length': '50', 'Connection': 'keep-alive', 'Set-Cookie': '__cfduid=d4d2298fb6e7fd06d3d3e9ba0bb0aee871593372823; expires=Tue, 28-Jul-20 19:33:43 GMT; path=/; domain=.kraken.zendesk.com; HttpOnly; SameSite=Lax, _help_center_session=b0pkZVBKVmZyMGdJZzNGMzJEQ2VEUHVPVUVWNy8xWmc1SnIzVzZla2xiWXR3OVAvZmpRbEhFT0lwbUpTeWRqWTNkczdWK3FZVXpYaGdZS0V3TFduWUE9PS0tUENLRElrSVJPNnFLSHpiTXJyYkltQT09--549b483f113291dfc300ba36249a0efc59a8565b; path=/; HttpOnly; Secure; SameSite=None, __cfruid=95b278f559ceeab826b6a85a821aa2ce478ca0cf-1593372823; path=/; domain=.kraken.zendesk.com; HttpOnly; Secure; SameSite=None', 'Retry-After': '17', 'X-Zendesk-Origin-Server': 'help-center-unicorn-6d96b967bb-2kzth', 'X-Request-Id': '5aa9d751c96fed1f-SJC', 'X-Runtime': '0.116356', 'Protocol': 'HTTP/1.0', 'CF-Cache-Status': 'BYPASS', 'cf-request-id': '039e04e7210000ed1f322ef200000001', 'Expect-CT': 'max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"', 'Vary': 'Accept-Encoding', 'Server': 'cloudflare', 'CF-RAY': '5aa9d751c96fed1f-SJC'} Number of allowed API requests per minute exceeded
2020-06-28 12:33:43,600 - freqtrade.exchange.exchange - WARNING - Unable to initialize markets. Reason: GET https://kraken.zendesk.com/api/v2/help_center/en-us/articles/205893708 429 Too Many Requests Number of allowed API requests per minute exceeded
2020-06-28 12:33:43,602 - freqtrade - ERROR - Cannot get markets. Reason: Markets were not loaded.
2020-06-28 12:33:43,602 - freqtrade.exchange.exchange - DEBUG - Exchange object destroyed, closing async loop
`

**My System:**

OS: Ubuntu 18.04
Python: 3.6.9
CCXT: 1.28.49
Branch: MASTER
Last Commit ID: bf53d037b53784c8e3263974ce81f0ac5219dced
FreqTrade: 2020.5

**Comment by xmatthias** [comment]
@Blackhawke updating ccxt will help, its a known kraken problem now fixed in ccxt.

`pip install -U ccxt`

**Comment by Blackhawke** [comment]
> @Blackhawke updating ccxt will help, its a known kraken problem now fixed in ccxt.
> 
> `pip install -U ccxt`

Yup! That seems to have fixed it. It's now downloading the trades.

My apologies for that. It's a new install, and the script downloaded and installed packages as needed, so I assumed it had installed the latest version of ccxt (1.28.49). And as the latest version of ccxt I saw here having the Kraken issue was .24.x, I concluded it wasn't the problem.

So, to be clear for others just starting out, ccxt 1.30.x or better is required if you're on Kraken?

**Comment by xmatthias** [comment]
well we'll be doing a release soon, so the pinned dependency will be updated.

for CCXT - whenever there are problems that originate from ccxt, updating almost always helps - they are VERY quick in fixing issues that arrise ...

**Comment by xmatthias** [comment]
I'll close this as we now have implemented API backoff ( #3531 ) in case of DDOS protection errors.

Now the recommendation is still to set rateLimit to a value that works for the exchange in question - as DDoS backoff can only save you to some extend - if you're a "reoccuring" offender the exchange will likely block the IP for the day...
